File: KSP/rltest.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <map>

#include "consts.h"
#include "craft.h"
#include "engines.h"
#include "random.h"
#include "body.h"
#include "nn.h"

//Constants
const double timestep = 0.1;

/*=== Simulation ===*/
struct SARGroup {
	std::vector<std::vector<double> > states;
	std::vector<std::vector<double> > actions;
	std::vector<double> rewards;
};

struct Result {
	bool valid;
	Craft state;
	double score;

	std::vector<SARGroup> critic_data;
};

Result simulate_ascent(Body body, Craft craft, double landing_alt, double orbit_alt, bool fuel_reverse, NN agent) {
	/*=== Constants ===*/
	double total_thrust = 0;
	for (auto const & iterator : craft.engines) {
		total_thrust += iterator.first.thrust * iterator.second;
	}

	/*=== Initial conditions ===*/
	double x = body.radius + landing_alt;
	double y = 0;
	double vx = 0;
	double vy = 2 * pi * x / body.rotational_period;

	/*=== NN inputs ===*/
	std::vector<double> nn_inputs = {
		0,
		0,
		0,
		0,
		body.radius * 1e-5, //Scaled to reduce magnitude for network
		std::log(body.grav_parameter),
		std::sqrt(body.grav_parameter / body.radius) * 1e-3,
		body.grav_parameter / body.radius / body.radius,
		2 * pi * body.radius / body.rotational_period,
		fuel_reverse ? -1.0 : 1.0
	};

	//Add engine inputs
	for (auto const & iterator : engines) {
		if (craft.engines.count(iterator.second) == 0) {
			nn_inputs.push_back(0);
			continue;
		}

		nn_inputs.push_back(craft.engines.at(iterator.second));
	}

	/*=== Recording ===*/
	SARGroup sar;

	/*=== Iterate ===*/
	double periapsis = 0;
	double sma = 0;
	while (periapsis < body.radius) {
		/*=== Fill NN inputs ===*/
		double dist = std::sqrt(x * x + y * y);
		double vh = (-y * vx + x * vy) / dist;
		double vv = ( x * vx + y * vy) / dist;

		nn_inputs[0] = dist;
		nn_inputs[1] = vh;
		nn_inputs[2] = vv;
		nn_inputs[3] = std::log(craft.mass);

		/*=== Find Action ===*/
		std::vector<double> action = agent.evaluate(nn_inputs);

		sar.states.push_back(nn_inputs);
		sar.actions.push_back(action);

		/*=== End conditions ===*/
		if (dist < body.radius) {
			//Crash
			sar.rewards.push_back(-1e6);
			break;
		}

		/*=== Calculate thrust ===*/
		double ah = action[0] / std::sqrt(action[0] * action[0] + action[1] * action[1]);
		double av = action[1] / std::sqrt(action[0] * action[0] + action[1] * action[1]);

		double thrust = total_thrust / (1 + std::exp(-action[2]));
		double accel = thrust / craft.mass;

		double tick_consumption = 0;
		for (auto const & iterator : craft.engines) {
			double group_thrust = iterator.first.thrust * iterator.second;

			if (group_thrust > thrust) {
				tick_consumption += thrust / iterator.first.isp / g0;
				break;
			}

			tick_consumption += group_thrust / iterator.first.isp / g0;
			thrust -= group_thrust;
		}

		/*=== Calculate accelerations ===*/
		ah *= accel;
		av *= accel;

		double ax = (-y * ah + x * av) / dist;
		double ay = ( x * ah + y * av) / dist;

		double grav = body.grav_parameter / dist / dist;
		double gx = -x * grav / dist;
		double gy = -y * grav / dist;

		/*=== Integrate ===*/
		vx += (ax + gx) * timestep;
		vy += (ay + gy) * timestep;
		x += vx * timestep;
		y += vy * timestep;

		craft.mass -= (fuel_reverse ? -1.0 : 1.0) * tick_consumption * timestep;

		/*=== Calculate periapsis ===*/
		sma = 1 / (2 / dist - (vx * vx + vy * vy) / body.grav_parameter);
		double eccentricity = std::sqrt(1 - (dist * dist * vh * vh) / (sma * body.grav_parameter));

		if (eccentricity > 1.0) {
			//Escape trajectory
			sar.rewards.push_back(-1e6);
			break;
		}

		periapsis = sma * (1 - eccentricity);

		/*=== Calculate rewards ===*/
		sar.rewards.push_back(-tick_consumption);
	}

	/*=== Use delta-v to raise/lower apoapsis ===*/
	double periapsis_vel =            std::sqrt(body.grav_parameter * (2 / periapsis - 1 / sma));
	double periapsis_vel_after_burn = std::sqrt(body.grav_parameter * (2 / periapsis - 2 / (periapsis + body.radius + orbit_alt)));

	craft.use_dv(periapsis_vel_after_burn - periapsis_vel);

	Result out;
	out.valid = (periapsis > body.radius);
	out.state = craft;
	out.score = craft.payload_fraction();

	out.critic_data.push_back(sar);

	return out;
}

Result simulate_landing(Body body, Craft craft, double landing_alt, double orbit_alt, NN agent) {
	/*=== Binary search to simulate landing ===*/
	double low = 0;
	double high = craft.mass;

	Craft state = craft;
	Result result;

	while (high - low > 1e-6) {
		double middle = (low + high) / 2;

		if (middle / craft.mass < 0.1) {
			return result;
		}

		state.mass = middle;

		result = simulate_ascent(body, state, landing_alt, orbit_alt, true, agent);

		if (!result.valid) {
			high = middle;
			continue;
		}

		if (result.state.mass > craft.mass) {
			high = middle;
			continue;
		}

		low = middle;
		continue;
	}

	/*=== Ascent ===*/
	state = result.state;
	Result ascent_result = simulate_ascent(body, state, landing_alt, orbit_alt, false, agent);

	ascent_result.critic_data.push_back(result.critic_data[0]);

	return ascent_result;
}

NN train_network(std::vector<int> layer_node_counts_agent) {
	/*=== Initialize NNs ===*/
	NN agent = get_random_nn(layer_node_counts_agent, Activation::ACTIVATION_RELU);

	layer_node_counts_agent[0] += 2;
	NN critic = get_random_nn(layer_node_counts_agent, Activation::ACTIVATION_RELU);

	/*=== Training ===*/
	for (int i = 0; i < 1000; i++) {
		/*=== Random body / craft ===*/
		Body body;

		body.radius = 850000 * std::pow(uniform(generator), 2.0);
		body.grav_parameter = std::abs(
			body.radius * 
			body.radius * 
			(1 + 0.2 * normal(generator)) * 
			((body.radius < 450000 ? 0.000009 : 0.000044) * (body.radius - 450000) + 4)
		);

		body.rotational_period = 150000 * std::pow(
			std::abs(std::log(uniform(generator))), 
			1.5
		);

		//Filters
		double surface_velocity = 2 * pi * body.radius / body.rotational_period;
		if (surface_velocity * surface_velocity * body.radius / body.grav_parameter > 0.25) {
			continue;
		}

		//Craft
		Craft craft;
		craft.mass = 100000;

		for (auto const & iterator : engines) {
			if (uniform(generator) > 0.1) continue;


		}

		std::cout << 
			body.radius << " " << 
			body.grav_parameter << " " << 
			body.rotational_period << " " << 
			body.grav_parameter / body.radius / body.radius << " " <<
			i << "\n";
	}

	return agent;
}

int main() {
	int num_engines = (int)engines.size();
	std::vector<int> layer_node_counts_agent = {num_engines + 10, 32, 32, 3};

	train_network(layer_node_counts_agent);
}

File: KSP/rltest.py
import tensorflow as tf
import keras
import numpy as np
from copy import deepcopy

from utils import *

#Constants
timestep = 0.1

#Engine sorting
engines_sorted = sorted([engine for engine in engine_info], key=lambda x: engine_info[x][0] * (1 + 1 / fuel_tank_info[engine_info[x][1]][1]), reverse=True)

#Simulation
def simulate_ascent(grav_parameter, radius, rotational_period, start_alt, fuel_reverse, craft_mass, craft_engines, agent, predictor):
	x = 0
	y = radius + start_alt
	vx = 2 * np.pi * y / rotational_period
	vy = 0
	mass = craft_mass

	total_thrust = sum([craft_engines[engine] * engine_info[engine][2] for engine in craft_engines])

	model_inputs = [craft_engines[engine] for engine in engine_info]
	model_inputs.append(grav_parameter / radius / radius)
	model_inputs.append(np.sqrt(grav_parameter / radius))
	model_inputs.append(radius)
	model_inputs.append(radius + start_alt)
	model_inputs.append(vx)
	model_inputs.append(vy)
	model_inputs.append(mass)

	engines = [engine for engine in engine_info]

	agent_inputs = []
	agent_outputs = []

	periapsis = 0
	time = 0
	while periapsis < radius:
		if time > 10000:
			break

		#Calculate acceleration
		dist = np.sqrt(x * x + y * y)
		if dist < radius:
			break

		vh = (-y * vx + x * vy) / dist
		vv = (x * vx + y * vy) / dist

		model_inputs[-4] = dist
		model_inputs[-3] = vh
		model_inputs[-2] = vv
		model_inputs[-1] = mass

		agent_inputs.append(deepcopy(model_inputs))

		throttle, th, tv = tuple(agent.predict(np.array(model_inputs).reshape((-1, 33)), verbose=0)[0])
		agent_outputs.append([throttle, th, tv])

		if throttle > 10:
			thrust = total_thrust
		elif throttle < -10:
			thrust = 0
		else:
			thrust = total_thrust / (1 + np.exp(-throttle))

		accel = thrust / mass
		ah = accel * th / np.sqrt(th * th + tv * tv)
		av = accel * tv / np.sqrt(th * th + tv * tv)

		ax = (-y * ah + x * av) / dist
		ay = (x * ah + y * av) / dist

		tick_consumption = {fuel: 0 for fuel in fuel_tank_info}
		for engine in engines_sorted:
			if engine not in craft_engines:
				continue

			group_thrust = craft_engines[engine] * engine_info[engine][2]

			if group_thrust > thrust:
				tick_consumption[engine_info[engine][1]] += thrust / engine_info[engine][0] / 9.80665
				break

			thrust -= group_thrust
			tick_consumption[engine_info[engine][1]] += group_thrust / engine_info[engine][0] / 9.80665

		#Gravity
		grav = grav_parameter / dist / dist
		gx = -x * grav / dist
		gy = -y * grav / dist

		consumption_mult = -1 if fuel_reverse else 1
		for consumption in tick_consumption.values():
			mass -= consumption_mult * consumption * timestep

		vx += (ax + gx) * timestep
		vy += (ay + gy) * timestep

		x += vx * timestep
		y += vy * timestep

		time += timestep

		sma = 1 / (2 / dist - (vx * vx + vy * vy))
		eccentricity = np.sqrt(1 - (dist * dist * vh * vh)/ sma / grav_parameter)

		periapsis = sma * (1 - eccentricity)

	if periapsis < radius:
		return [periapsis/radius, agent_inputs, agent_outputs]
	return [1, agent_inputs, agent_outputs]

def simulate_landing(grav_parameter, radius, rotational_period, land_alt, orbit_alt, craft_mass, craft_engines, agent, predictor):
	low = 0
	high = craft_mass

	while high - low > 1e-3:
		middle = (low + high) / 2

		result = simulate_ascent(grav_parameter, radius, rotational_period, land_alt, True, middle, craft_engines, agent, predictor)

		print([a + b for a, b in zip(result[1], result[2])])

		low = middle


#Model definitions
def get_agent():
	inputs = keras.Input(shape=(len(engine_info) + 7,), name="state")
	x1 = keras.layers.Dense(32, activation="relu")(inputs)
	x2 = keras.layers.Dense(32, activation="relu")(x1)
	outputs = keras.layers.Dense(3, name="action")(x2)

	model = keras.Model(inputs=inputs, outputs=outputs)
	return model
	
def get_predictor():
	inputs = keras.Input(shape=(len(engine_info) + 10,), name="stateaction")
	x1 = keras.layers.Dense(32, activation="relu")(inputs)
	x2 = keras.layers.Dense(32, activation="relu")(x1)
	outputs = keras.layers.Dense(1, name="reward")(x2)

	model = keras.Model(inputs=inputs, outputs=outputs)
	return model

#Get networks
agent = get_agent()
predictor = get_predictor()

agent.summary()

#Training classes
optimizer = keras.optimizers.SGD(learning_rate=1e-3)
loss = keras.losses.MeanSquaredError()

#Train
for i in range(1000):
	radius = 2e5
	grav_parameter = 65138397520.7807
	rotational_period = 138984.376574476

	craft_engines = {engine: 0 for engine in engine_info}
	craft_engines["Vector"] = 1

	simulate_landing(grav_parameter, radius, rotational_period, 1000, 10000, 1e5, craft_engines, agent, predictor)